Pass timeout through download_files_in_parallel

MSeifert04 opened the pull request:
Previously the `timeout` argument was simply ignored.

MSeifert04 commented on astropy/utils/data.py:
```diff
@@ -1159,15 +1159,11 @@ def download_files_in_parallel(urls, cache=False, show_progress=True,
     else:
         progress = io.BytesIO()
 
-    if timeout is None:
```
I think that was just "dead code", the `timeout is None` case was already covered a few lines above.

MSeifert04 commented on astropy/utils/data.py:
```diff
@@ -986,10 +986,6 @@ def download_file(remote_url, cache=False, show_progress=True, timeout=None):
 
     missing_cache = False
 
-    if timeout is None:
```
same here

pllim replied:
Good catch!

MSeifert04 commented:
Not sure whom to assign or request a review from. @pllim @bsipocz?

bsipocz commented:
Either of us is OK, but @pllim is the lead of utils :)

pllim commented on astropy/utils/data.py:
```diff
@@ -1118,7 +1114,7 @@ def is_url_in_cache(url_key):
 
 
 def _do_download_files_in_parallel(args):
```
Do we even still need this? Can `map` just call `download_file` directly now?

MSeifert04 replied:
I don't think so, it still requires the unpacking.

pllim replied:
Maybe there is a way to unpack args without using a dummy in-between function? Anyway, that should not block the merge.

MSeifert04 replied:
Well, there is `starmap` (however it's more like `map` than the currently used `imap`) and maybe something in `concurrent.futures`, however I didn't want to rewrite `ProcessBar`... and it wouldn't be a clean backport to 2.0.3 because of the Python 2 support. Also it would mix a straightforward bugfix to something like an enhancement.

MSeifert04 replied:
A decorator would be easy to write (but again that would make it an enhancement rather than a bugfix). Something like (shameless self-promotion ahead: [`packed`](https://iteration-utilities.readthedocs.io/en/stable/generated/packed.html)).

bsipocz replied:
hey, you don't have a link to the source repo at any prominent place?

MSeifert04 replied:
It's essentially just:

```
class packed(object):
  __slots__ = ('func', )
  def __init__(self, func):
    self.func = func
  def __call__(self, args, **kwargs):
    return self.func(*args, **kwargs)
```

but [written as C extension](https://github.com/MSeifert04/iteration_utilities/blob/master/src/packed.c) (with some minor additional stuff).

bsipocz replied:
(yes, I found it, I just meant that from the docs of `iteration_utilities` one has to dig down to find a link to github ;) )

MSeifert04 replied:
Seems like I, finally, have a reason to look into "how to create a footer/header" with sphinx. 😄

bsipocz replied:
you can opt in the annoying "edit on github" plugin, that's the easiest ;)

MSeifert04 replied:
I'm almost certain that it won't work because of the C extension (like the `[source]` "button").

pllim replied:
There is a way to attach docstring into your C-extension. See [astropy/wcs/docstrings.py](https://github.com/astropy/astropy/blob/master/astropy/wcs/docstrings.py).

MSeifert04 replied:
I don't doubt that, but the "edit on github" and "source" buttons always refer to the Python function or don't display.

But I'm getting side-tracked. 😄

pllim replied:
That is true. It does not point straight to the C code.

pllim commented:
Unfortunately, I can't find any test for the parallel stuff in the `utils` test suite, nor used in other sub-packages being tested. So tests passing here don't really test your changes. 🤔

MSeifert04 commented:
So this needs a test?

pllim commented:
> So this needs a test?

Ideally, yes. But is it possible to do in a CI is another question I cannot answer? I don't know if there was ever a reason for not having a test.

bsipocz commented:
@pllim - I remember a few cases where adding tests would have been a big nightmare hack and the gain was just nitpicking.

pllim commented:
At the very least, if adding test is not possible, we should confirm that this runs as intended ™️ locally.

MSeifert04 commented:
I'll try to make a test for it, but I don't think testing the timeout is feasible. Except you know any online resource that reliably takes more than a specified amount to load.

pllim commented:
Will setting `timeout=0` ensures a timeout? Anyway, yeah, we don't have to go overboard on testing this. Thanks for looking into it!

bsipocz commented:
It's fine not to cover timeout, but @pllim mentions above that the whole parallel donwload stuff is untested?

pllim commented:
> the whole parallel donwload stuff is untested?

As far as I could `grep`, I couldn't find any. I wonder why it was written in the first place. It is not used anywhere in the core, nor in Cone Search that I just moved out.

bsipocz commented:
probably it looked useful at the time of writing? (mostly untouched since 201301)

MSeifert04 commented:
I added a test and rebased this. However it seems like there's something fishy going on. If `cache=False` then the test fails on my computer (probably because the files are cleaned when the process ends ... no time to investigate that now but could be another bug in hiding).

I added `(source='astropy')` right now so that it will be run in the mandatory test section. However I need to remove that again before it's merged.

MSeifert04 commented on astropy/utils/data.py:
```diff
@@ -18,6 +18,8 @@
 import sys
 import time
 import urllib
+import urllib.error
```
When I was running the file with the `-t` option it needed these imports.

drdavella replied:
Thanks for the comment; this is independent verification that #6668 is actually necessary (and that I'm not just imagining things).

MSeifert04 commented:
Oups, I should have checked locally with `--remote-data=astropy` 😓

MSeifert04 commented:
Okay, the one remote-data travis run worked. The other just had to fail with a TimeOut... I restarted it.

Is the test basically okay like that?

MSeifert04 commented on astropy/utils/tests/test_data.py:
```diff
@@ -46,6 +46,17 @@ def test_download_nocache():
+        ['http://data.astropy.org',
+         'https://astropy.stsci.edu/data/intersphinx/README'],
+        cache=True)
+    assert all([os.path.isfile(f) for f in fnout]), fnout
```
ah, I see I've forgotten to remove the "assert failure message" here (the trailing `, fnout`. Can that stay in or should it be removed (I've never seen it used in other tests).

pllim replied:
If the value of `fnout` is useful for debugging, no harm leaving it in. But the test will never fail again, right, so this is a non-issue??? 😉

MSeifert04 replied:
I was using it when `cache=False` kept failing to see where the files were "supposed" to be.

I guess this test will fail a lot if we use it with `remote_data('astropy')` (timeout ... yay 🎉) but not in the assert 😄

pllim commented:
Ops, there is conflict in the change log

MSeifert04 commented:
Is it otherwise ready-to-be-merged?

pllim commented:
This LGTM and there are no objections from deputies, so yes.

MSeifert04 commented:
okay, I fixed the conflict and made the test `remote_data` without `'astropy'` so it doesn't execute too many times.

pllim commented on astropy/utils/tests/test_data.py:
```diff
@@ -47,6 +47,17 @@ def test_download_nocache():
+    fnout = download_files_in_parallel(
+        ['http://data.astropy.org',
+         'https://astropy.stsci.edu/data/intersphinx/README'],
+        cache=True)
```
We should note #6662 here on why `cache=True`.

MSeifert04 commented:
Passed the remote-data test (caution long log): https://travis-ci.org/astropy/astropy/jobs/283826808#L2763

MSeifert04 commented:
CI passed and it has the "merge-when-ci-passes" label but I don't like to merge my own PR.

drdavella commented:
Just for the record, I think this was my first time merging something to astropy core.

pllim commented:
@drdavella with great power comes great responsibility! 🕸

bsipocz commented:
@drdavella - wow, magic powers ;)

bsipocz commented:
Due to the urllib imports this cannot be backported as is, but with a simple workaround using six. I'll commit that directly to the 2.0.x branch.
